src/test/mir-opt/const_prop/read_immutable_static.rs

```diff
@@ -0,0 +1,27 @@
+static FOO: u8 = 2;
+
+fn main() {
+    let x = FOO + FOO;
+}
+
+// END RUST SOURCE
+// START rustc.main.ConstProp.before.mir
+//  bb0: {
+//      ...
+//      _2 = (FOO: u8);
+//      ...
+//      _3 = (FOO: u8);
+//      _1 = Add(move _2, move _3);
+//      ...
+//  }
+// END rustc.main.ConstProp.before.mir
+// START rustc.main.ConstProp.after.mir
+//  bb0: {
+//      ...
+//      _2 = const 2u8;
+//      ...
+//      _3 = const 2u8;
+//      _1 = Add(move _2, move _3);
```

Comment:
huh, why weren't these propped into the add?

Reply:
I think we only have handling for prop'ing operands into the `Assert` and `SwitchInt` terminators currently. For statements, if can eval the statement and assign it to the correct place, then we'll replace it. We can eval the statement but we can't assign it to the place because it's a local. I believe if I change the test to do `(FOO + FOO) + 0` we'll see the inner `Add()` turned into `const 4u8`.

Reply:
Ah, makes sense. It would be totally a legal change though, since we aren't setting the internal value for locals, so we don't prop out of the local, just into the assignment.

Reply:
I agree. There's other limitations around here

https://github.com/rust-lang/rust/blob/c7bc0bf82faf9718fa1e59a38f5aab308299ba7d/src/librustc_mir/transform/const_prop.rs#L603

and here

https://github.com/rust-lang/rust/blob/c7bc0bf82faf9718fa1e59a38f5aab308299ba7d/src/librustc_mir/transform/const_prop.rs#L628

that would be great to improve as well. More work for another day 🙂